File: LeetCode/C/189-array-旋转数组.cpp
/*
 * @lc app=leetcode.cn id=189 lang=cpp
 *
 * [189] 旋转数组
 */
#include "main.h"

// @lc code=start
class Solution {
  public:
    void rotate(vector<int> &nums, int k) {
        int n = nums.size();
        k %= n;
        reverse(nums, 0, n);
        reverse(nums, 0, k);
        reverse(nums, k, n);
    }

  private:
    void reverse(vector<int> &nums, int left, int right) {
        while (left < right)
            swap(nums[left++], nums[--right]);
    }
};
// @lc code=end

File: LeetCode/Go/189-array-旋转数组.go
/*
 * @lc app=leetcode.cn id=189 lang=golang
 *
 * [189] 旋转数组
 */
package leetcode

// @lc code=start
func rotate(nums []int, k int) {
	reverse := func(l, r int) {
		for l < r {
			r--
			nums[l], nums[r] = nums[r], nums[l]
			l++
		}
	}
	n := len(nums)
	k %= n
	reverse(0, n)
	reverse(0, k)
	reverse(k, n)
}

// @lc code=end
